Subgradients.thy

```diff
@@ -0,0 +1,371 @@
+theory Subgradients
+  imports "RLM_stable"
+begin
+
+
+definition subgradient :: "  ('a::euclidean_space \<Rightarrow> real) \<Rightarrow> 'a::euclidean_space set \<Rightarrow> 
+          'a::euclidean_space \<Rightarrow> 'a::euclidean_space \<Rightarrow> bool " where
+  "subgradient f H' w v = (w \<in> H' \<and> ( \<forall> u \<in> H'. f (u) \<ge> f(w) + (inner (u-w) v)))"
+
+definition diff_set :: "('a::euclidean_space \<Rightarrow> real) \<Rightarrow> 'a::euclidean_space set \<Rightarrow> 
+          'a::euclidean_space \<Rightarrow> 'a::euclidean_space set" where
+  "diff_set f H' w = {h. subgradient f H' w h}"
+
+fun hinge_loss :: "('a:: euclidean_space  \<Rightarrow> ('a \<times> real) \<Rightarrow> real)" where
+  "hinge_loss w pair = (max 0 (1 - (snd pair) * inner w (fst pair)))" 
+
+
+text\<open>Don't need to prove it, if we show it is true for the hinge loss\<close>
+lemma convex_has_subgradient:
+  fixes f :: "'a::euclidean_space \<Rightarrow> real"
+  fixes A :: "'a::euclidean_space set"
+  fixes \<rho> :: "real"
+  assumes "\<rho> \<ge> 0"
+  assumes " open A"
+  assumes "convex A"
+  shows "convex_on A f \<longleftrightarrow> (\<forall> w \<in> A. \<exists> v \<in> A. \<forall> u \<in> A. f (u) \<ge> f(w) + (inner (u-w) v)) "
+  sorry
+
+
+
+
+lemma 147 :
```

Comment:
can you find a better name for it, and add comment that it relates to lemma 14.7?